src/core/utils/components/VirtualScroll.tsx

```diff
@@ -0,0 +1,231 @@
+    const arr = [];
+    const endIndex = Math.min(
+      childrenLength,
+      startNode + visibleNodeCount + buffer,
+    );
+    for (let i = startNode; i < endIndex; i++) {
+      arr.push(itemRenderer(i));
+    }
+    return arr;
+  }, [childrenLength, itemRenderer, buffer, startNode, visibleNodeCount]);
+
+  // Get child height when children available
+  React.useLayoutEffect(() => {
+    if (!parentRef.current || !visibleChildren.length) {
+      return;
+    }
+
+    const firstChild = parentRef.current.children.item(0) as HTMLElement;
+    childHeight.current = Number(getElementHeight(firstChild).toFixed(2));
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [visibleChildren.length]);
+
+  const onScroll = React.useCallback(
+    (e: Event) => {
+      const target =
+        scrollContainer.current !== document.body
+          ? (e.target as HTMLElement)
+          : (document.scrollingElement as HTMLElement);
+      if (!target) {
+        return;
+      }
+      const start = getNumberOfNodesInHeight(
+        childHeight.current,
+        target.scrollTop,
+      );
+      setStartNode(start);
+      setVisibleNodeCount(
+        getVisibleNodeCount(childHeight.current, start, childrenLength, target),
+      );
+
+      if (!parentRef.current) {
+        return;
+      }
+      parentRef.current.style.transform = `translateY(${getTranslateValue(
+        childHeight.current,
+        start,
+      )}px)`;
+    },
+    [childrenLength],
+  );
+
+  const removeScrollListener = React.useCallback(() => {
+    if (!onScrollRef.current) {
+      return;
+    }
+    !scrollContainer.current || scrollContainer.current === document.body
+      ? document.removeEventListener('scroll', onScrollRef.current)
+      : scrollContainer.current.removeEventListener(
+          'scroll',
+          onScrollRef.current,
+        );
+  }, []);
+
+  // Add event listener to the scrollable container.
+  React.useLayoutEffect(() => {
+    removeScrollListener();
+    onScrollRef.current = onScroll;
+    if (!scrollContainer.current || scrollContainer.current === document.body) {
+      document.addEventListener('scroll', onScroll);
+    } else {
+      scrollContainer.current.addEventListener('scroll', onScroll);
+    }
+    return removeScrollListener;
+  }, [onScroll, removeScrollListener]);
+
+  React.useLayoutEffect(() => {
+    const scrollableContainer = scrollContainer.current ?? document.body;
+    const top = scrollableContainer.scrollTop;
+    const start = getNumberOfNodesInHeight(childHeight.current, top);
+    setStartNode(start);
+    setVisibleNodeCount(
+      getVisibleNodeCount(
+        childHeight.current,
+        start,
+        childrenLength,
+        scrollableContainer,
+      ),
+    );
+
+    if (!parentRef.current) {
+      return;
+    }
+    parentRef.current.style.transform = `translateY(${getTranslateValue(
+      childHeight.current,
+      start,
+    )}px)`;
+  }, [childrenLength]);
+
```

Comment:
I think you might need to recalculate stuff on resize.

Reply:
so no, I think?